autofill/autofill-impl/src/test/java/com/duckduckgo/autofill/impl/importing/DefaultDomainNameNormalizerTest.kt

```diff
@@ -0,0 +1,71 @@
+package com.duckduckgo.autofill.impl.importing
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.duckduckgo.autofill.api.domain.app.LoginCredentials
+import com.duckduckgo.autofill.impl.encoding.UrlUnicodeNormalizerImpl
+import com.duckduckgo.autofill.impl.urlmatcher.AutofillDomainNameUrlMatcher
+import kotlinx.coroutines.test.runTest
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class DefaultDomainNameNormalizerTest {
+
+    private val testee = DefaultDomainNameNormalizer(AutofillDomainNameUrlMatcher(UrlUnicodeNormalizerImpl()))
+
+    @Test
+    fun whenEmptyInputThenEmptyOutput() = runTest {
+        val input = emptyList<LoginCredentials>()
+        val output = testee.normalizeDomains(input)
+        assertTrue(output.isEmpty())
+    }
+
+    @Test
+    fun whenInputDomainAlreadyNormalizedThenIncludedInOutput() = runTest {
+        val input = listOf(creds(domain = "example.com"))
+        val output = testee.normalizeDomains(input)
+        assertEquals(1, output.size)
+        assertEquals(input.first(), output.first())
+    }
+
+    @Test
+    fun whenInputDomainNotAlreadyNormalizedThenNormalizedAndIncludedInOutput() = runTest {
+        val input = listOf(creds(domain = "https://example.com/foo/bar"))
+        val output = testee.normalizeDomains(input)
+        assertEquals(1, output.size)
+        assertEquals(input.first().copy(domain = "example.com"), output.first())
+    }
+
+    @Test
+    fun whenInputDomainIsNullThenNormalizedToNullDomain() = runTest {
+        val input = listOf(creds(domain = null))
+        val output = testee.normalizeDomains(input)
+        assertEquals(1, output.size)
+        assertEquals(null, output.first().domain)
+    }
+
+    @Test
+    fun whenDomainCannotBeNormalizedThenIsIncludedUnmodified() = runTest {
```

Comment:
I'm not sure about this one 😅. If we don't know how to normalize the input, we just return the input. But then we store it as domain attribute...but that can be anything. 🤔

You have more context than me on Autofill, but what are the side effects of not having a domain in the domain attribute? I assume we are not validating the user manually enters something valid either...so at this point, I wonder why do we even normalize, or why do we even filter the "android://..." objects... 

Short term, just to consider if this is the expected behavior, and then we can chat about the class fields without blocking the changes here as it seems a more broad discussion.

Reply:
This behavior mirrors what we do when saving credentials. If we can, we normalize the URLs when we encounter them, or when the user enters them manually (a product decision from way back). If the URL is unrecongizable as a URL and therefore we can't normalize it (when saving from autofill or from a user entering manually), we save it unaltered.

Reply:
>what are the side effects of not having a domain in the domain attribute

if it's unrecognizable, we won't be able to match on it and it won't be offered as an option in the autofill dialogs but will still be stored in the full credential list